archive/100-theorems-list/88_derangements_formula.lean

```diff
@@ -0,0 +1,468 @@
+
+/-- For any fixed `a b : α`, the permutations that send `a` to `b` are equivalent to permutations
+    away from `a`, by omitting `a` and sending `f⁻¹ a` to `b` instead. -/
+def perm_splice_out (a b : α) : { f : perm α | f a = b } ≃ perm ({a}ᶜ : set α) :=
+begin
+  let just_a : set α := {a},
+
+  transitivity {f : perm α // ∀ x : just_a, f x = equiv.refl just_a x},
+  { -- we want to modify each permutation by composing it with (swap a and b)
+    let compose_with_swap : perm (perm α) := equiv_compose (equiv.swap a b),
+
+    apply equiv.trans (subtype_equiv_of_subtype' compose_with_swap) _,
+    apply subtype_equiv_right,
+    simp [compose_with_swap, equiv_compose],
+    simp [apply_eq_iff_eq_symm_apply] },
+  { apply equiv.set.compl }
+end
+
+lemma perm_splice_out_eq (a b : α) (f : perm α) (fa_eq_b : f a = b) (x : α) (hx : x ≠ a) :
+  ((perm_splice_out a b ⟨f, fa_eq_b⟩ ⟨x, hx⟩) : α) = (swap a b) (f x) :=
+begin
+  simp [perm_splice_out, equiv.set.compl,
+        subtype_equiv_of_subtype', subtype_equiv_of_subtype,
+        equiv_compose],
+end
+
+lemma splice_derangement (a b : α) (hab : a ≠ b) (f : perm α) (fa_eq_b : f a = b) :
+  f b ≠ a ∧ is_derangement f ↔ (perm_splice_out a b) ⟨f, fa_eq_b⟩ ∈ derangements ({a}ᶜ : set α) :=
+begin
+  -- again, easier to show that fixed points lead to other fixed points
+  rw ← not_iff_not,
+  simp [derangements, is_derangement],
+  rw imp_iff_not_or,
+  simp,
+
+  split,
+  { -- if f has a fixed point, then so does (perm_splice_out a b) f.
+    -- also, if f b = a, then b is a fixed point
+    intro hyp,
+    cases hyp with fb_eq_a x_fixed,
+
+    { use [b, hab.symm],
+      simp [subtype.ext_iff_val, perm_splice_out_eq],
+      rw apply_eq_iff_eq_symm_apply,
+      simpa },
+
+    { rcases x_fixed with ⟨x, x_fixed⟩,
+
+      have x_ne_a : x ≠ a,
+      { intro contra,
+        rw contra at x_fixed,
+        cc },
+
+      have x_ne_b : x ≠ b,
+      { intro contra,
+        rw contra at x_fixed,
+        have : f a = f b, by cc,
+        exact hab (f.injective this) },
+
+      use [x, x_ne_a],
+      simp [subtype.ext_iff_val, perm_splice_out_eq],
+      rw apply_eq_iff_eq_symm_apply,
+      simp [x_fixed, swap_apply_of_ne_of_ne x_ne_a x_ne_b] } },
+  { -- if (perm_splice_out a b) f has a fixed point, either it came from a pre-existing
+    -- one, or it's b
+    rintro ⟨x, x_ne_a, x_fixed⟩,
+    simp [subtype.ext_iff_val, perm_splice_out_eq] at x_fixed,
+    apply_fun (swap a b) at x_fixed,
+    simp at x_fixed,
+
+    by_cases x_vs_b : x = b,
+    { left,
+      simp [x_vs_b] at x_fixed,
+      assumption },
+    { right,
+      use x,
+      rw swap_apply_of_ne_of_ne x_ne_a x_vs_b at x_fixed,
+      assumption }
+  }
+end
+
+end splicing
+
+
+section recursion
+
+-- TODO: these are just here to help the elaborator. I would like to eliminate
+-- them in favor of `:` coercions, but I keep getting errors about it. :\
+def everything_but {α : Type*} (a : α) : set α := {a}ᶜ
```

Comment:
What's wrong with just writing out `({a}ᶜ : set α)` each time? It's about the same length anyway.

Reply:
i had some issues with not being able to infer the right typeclass (something something `has_coe_to something`). lemme see if i can make it happen again

Reply:
Actually, I've had issues with \compl too before; I often can't write it in front of a set and have to write `.compl`, which is defeq. Henry, does it work if you use `.compl`?

Reply:
whoop, forgot to resolve this, but no, i got the same issue with "can't infer typeclass" :(